jdaviz/core/template_mixin.py

```diff
@@ -136,92 +136,164 @@ class SpectralSubsetSelect(BasePluginComponent):
+        """
+        Parameters
+        ----------
+        plugin
+            the parent plugin object
+        items : str
+            the name of the items traitlet defined in ``plugin``
+        selected : str
+            the name of the selected traitlet defined in ``plugin``
+        selected_has_subregions: str
+            the name of the selected_has_subregions traitlet defined in ``plugin``, optional
+        viewer_refs : list
+            the reference names of the viewer to extract the subregion.  If not provided or None,
+            will loop through all references.
+        default_text : str or None
+            the text to show for no selection.  If not provided or None, no entry will be provided
+            in the dropdown for no selection.
+        allowed_type : str or None
+            whether to filter to 'spatial' or 'spectral' types of subsets.  If not provided or None,
+            will include both entries.
+        """
         super().__init__(plugin,
                          items=items,
                          selected=selected,
                          selected_has_subregions=selected_has_subregions)
-        self.hub.subscribe(self, SubsetUpdateMessage, handler=self._update_spectral_subset)
-        self.hub.subscribe(self, SubsetDeleteMessage, handler=self._delete_spectral_subset)
+
+        self._viewer_refs = viewer_refs
+        self._default_text = default_text
+        if allowed_type not in [None, 'spatial', 'spectral']:
+            raise ValueError("allowed_type must be None, 'spatial', or 'spectral'")
+        self._allowed_type = allowed_type
+
+        # set default values for traitlets
+        if default_text is not None:
+            self.items = [{"label": default_text}]
+            self.selected = default_text
+        if selected_has_subregions is not None:
+            self.selected_has_subregions = False
+
+        self.hub.subscribe(self, SubsetUpdateMessage,
+                           handler=lambda msg: self._update_subset(msg.subset, msg.attribute))
+        self.hub.subscribe(self, SubsetDeleteMessage,
+                           handler=lambda msg: self._delete_subset(msg.subset))
         self.add_observe(selected, self._selected_changed)
 
-    def _subset_to_dict(self, subset):
-        # find layer artist in default spectrum-viewer
-        for layer in self.spectrum_viewer.layers:
-            if layer.layer.label == subset.label:
-                color = layer.state.color
-                break
+        # intialize any subsets that have already been created
+        for lyr in self.app.data_collection.subset_groups:
+            self._update_subset(lyr)
+
+    @property
+    def viewer_refs(self):
+        if self._viewer_refs is None:
+            return [ref for ref in self.app.get_viewer_reference_names() if ref is not None]
+        return self._viewer_refs
+
+    @property
+    def viewers(self):
+        return [self.app.get_viewer(ref) for ref in self.viewer_refs]
```

Comment:
this is likely the cause of #1217 - when `_subset_to_dict` loops through the viewers by ref in imviz, only the original viewer has a reference.  The subset still exists in the original viewer, so it is able to get the label and type, but the color must not be set when the message is received.